Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,234 @@
+    if (match.minor_number_allocation_type == MinorNumberAllocationType::SequentialLimited)
+        allocation_map_size = match.minor_number_range_size;
+    auto bitmap = TRY(Bitmap::create(allocation_map_size, false));
+    auto node = TRY(adopt_nonnull_own_or_enomem(new (nothrow) DeviceNodeFamily(move(bitmap),
+        match.family_type_literal,
+        block_device ? DeviceNodeFamily::Type::BlockDevice : DeviceNodeFamily::Type::CharacterDevice,
+        major_number,
+        minor_number)));
+    TRY(m_device_node_families.try_append(move(node)));
+
+    return m_device_node_families.last().ptr();
+}
+
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
+        allocation_index = allocation_index / 26;
+    } while (allocation_index > 0);
+    return base_string;
+}
+
+static ErrorOr<String> build_suffix_with_numbers(size_t allocation_index)
+{
+    return String::number(allocation_index);
+}
+
+static ErrorOr<void> prepare_permissions_after_populating_devtmpfs(StringView path, DeviceEventLoop::device_node_match const& match)
+{
+    if (match.permission_group.is_null())
+        return {};
+    auto group = TRY(Core::System::getgrnam(match.permission_group));
+    VERIFY(group.has_value());
+    TRY(Core::System::chown(path, 0, group.value().gr_gid));
+    TRY(Core::System::endgrent());
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::register_new_device(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    auto* possible_match = device_node_family_to_match_type(block_device, major_number, minor_number);
+    if (!possible_match)
+        return {};
+    auto& match = *possible_match;
+    auto& device_node_family = *TRY(find_or_register_new_device_node_family(match, block_device, major_number, minor_number));
+
+    static constexpr StringView devtmpfs_base_path = "/dev/"sv;
+    auto path_pattern = TRY(String::from_utf8(match.path_pattern));
+    auto& allocation_map = device_node_family.devices_symbol_suffix_allocation_map();
+    auto possible_allocated_suffix_index = allocation_map.find_first_unset();
+    if (!possible_allocated_suffix_index.has_value()) {
+        // FIXME: Make the allocation map bigger?
+        return Error::from_errno(ERANGE);
+    }
+    auto allocated_suffix_index = possible_allocated_suffix_index.release_value();
+
+    auto path = TRY(String::from_utf8(path_pattern));
+    if (match.path_pattern.contains("%d"sv)) {
+        auto replacement = TRY(build_suffix_with_numbers(allocated_suffix_index));
+        path = TRY(path.replace("%d"sv, replacement, ReplaceMode::All));
+    }
+    if (match.path_pattern.contains("%c"sv)) {
+        auto replacement = TRY(build_suffix_with_letters(allocated_suffix_index));
+        path = TRY(path.replace("%c"sv, replacement, ReplaceMode::All));
+    }
```

Comment:
It's basically the same like what `printf()` does, so I can't really understand what is the problem. Maybe adding a comment will solve this?

Reply:
However, especially %c is not really comparable to printf since it can print more than one character. The documentation must then be with the path pattern variable.